File: v2/Robot/WebDriverWrapper/src/com/webdriver/wrapper/WebDriverImpl.java
package com.webdriver.wrapper;

import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.util.ArrayList;
import java.util.List;

class WebDriverImpl implements WebDriver {
    private final org.openqa.selenium.WebDriver webDriver;
    private final JavascriptExecutor javascriptExecutor;
    private final WebDriverWait webDriverWait;

    WebDriverImpl(org.openqa.selenium.WebDriver webDriver, int waitTimeoutInSeconds) {
        this.webDriver = webDriver;
        this.webDriverWait = new WebDriverWait(this.webDriver, waitTimeoutInSeconds);

        this.javascriptExecutor = (JavascriptExecutor)this.webDriver;
    }

    @Override
    public void navigate(String url, boolean waitCompletion) {
        webDriver.get(url);

        if (waitCompletion) {
            waitPageLoadCompleted();
        }
    }

    @Override
    public List<WebElement> getElementsByClassName(String className) {
        List<WebElement> resultElements = new ArrayList<>();

        String javascriptBody = String.format("return document.getElementsByClassName('%s');", className);
        List<org.openqa.selenium.WebElement> webElements = (List<org.openqa.selenium.WebElement>)javascriptExecutor.executeScript(javascriptBody);
        for (org.openqa.selenium.WebElement webElement : webElements) {
            resultElements.add(new WebElementImpl(webDriver, webElement, javascriptExecutor, webDriverWait));
        }

        return resultElements;
    }

    @Override
    public WebElement getElementById(String elementId) {
        String javascriptBody = String.format("return document.getElementById('%s');", elementId);
        org.openqa.selenium.WebElement webElement = (org.openqa.selenium.WebElement)javascriptExecutor.executeScript(javascriptBody);

        return new WebElementImpl(webDriver, webElement, javascriptExecutor, webDriverWait);
    }

    @Override
    public void waitPageLoadCompleted() {
        webDriverWait.until(new PageCompletedWaitCondition(javascriptExecutor));
    }

    @Override
    public void quit() {
        webDriver.quit();
    }

    public Object executeJavaScript(String scriptBody) {
        return javascriptExecutor.executeScript(scriptBody);
    }
}

File: v2/dotNET/WPF/RobotWrapper/Wrapper/Internal/WebDriverImpl.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Threading;
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;
using ISeleniumWebDriver = OpenQA.Selenium.IWebDriver;
using ISeleniumWebElement = OpenQA.Selenium.IWebElement;

namespace WebDriver.Wrapper.Internal
{
    internal class WebDriverImpl : IWebDriver
    {
        private readonly ISeleniumWebDriver _webDriver;
        private readonly IJavaScriptExecutor _javaScriptExecutor;
        private readonly WebDriverWait _webDriverWait;

        public WebDriverImpl(ISeleniumWebDriver webDriver, TimeSpan timeout)
        {
            _webDriver = webDriver;
            _javaScriptExecutor = _webDriver as IJavaScriptExecutor;
            _webDriverWait = new WebDriverWait(_webDriver, timeout);
        }

        public void Navigate(string url, bool waitCompletion)
        {
            _webDriver.Url = url;

            if (waitCompletion)
            {
                WaitPageLoadCompleted();
            }
        }

        public IEnumerable<IWebElement> GetElementsByClassName(string className)
        {
            var resultElements = new List<IWebElement>();

            var javascriptBody = string.Format(
                CultureInfo.InvariantCulture,
                "return document.getElementsByClassName('{0}');", 
                className);

            var webElements = ExecuteJavaScript<IEnumerable<object>>(javascriptBody);

            foreach (var webElement in webElements)
            {
                resultElements.Add(new WebElementImpl((ISeleniumWebElement)webElement, _javaScriptExecutor));
            }

            return resultElements;
        }

        public IWebElement GetElementById(string elementId)
        {
            var javascriptBody = string.Format(
                CultureInfo.InvariantCulture,
                "return document.getElementById('{0}');", 
                elementId);

            var webElement = ExecuteJavaScript<ISeleniumWebElement>(javascriptBody);

            return new WebElementImpl(webElement, _javaScriptExecutor);
        }

        public IEnumerable<IWebElement> GetElementsByClassName(string className, TimeSpan timeout, int retryCount)
        {
            var resultElements = new List<IWebElement>();

            var javascriptBody = string.Format(
                CultureInfo.InvariantCulture,
                "return document.getElementsByClassName('{0}');",
                className);

            var webElements = ExecuteJavaScript<IEnumerable<object>>(javascriptBody);

            while (retryCount-- > 0 && !webElements.Any())
            {
                Thread.Sleep(timeout);
            }

            foreach (var webElement in webElements)
            {
                resultElements.Add(new WebElementImpl((ISeleniumWebElement)webElement, _javaScriptExecutor));
            }

            return resultElements;
        }

        public IWebElement GetElementById(string elementId, TimeSpan timeout, int retryCount)
        {
            var javascriptBody = string.Format(
                CultureInfo.InvariantCulture,
                "return document.getElementById('{0}');",
                elementId);

            var webElement = ExecuteJavaScript<ISeleniumWebElement>(javascriptBody);

            while (retryCount-- > 0 && webElement == null)
            {
                Thread.Sleep(timeout);
            }

            return new WebElementImpl(webElement, _javaScriptExecutor);
        }

        public void WaitPageLoadCompleted()
        {
            _webDriverWait.Until(IsPageLoaded);
        }

        public void Quit()
        {
            _webDriver.Quit();
        }

        public object ExecuteJavaScript(string scriptBody)
        {
            return _javaScriptExecutor.ExecuteScript(scriptBody);
        }

        public TResult ExecuteJavaScript<TResult>(string scriptBody)
        {
            return (TResult)_javaScriptExecutor.ExecuteScript(scriptBody);
        }

        private bool IsPageLoaded(ISeleniumWebDriver webDriver)
        {
            var jQueryType = (string)_javaScriptExecutor.ExecuteScript("return typeof jQuery");
            if (jQueryType.Equals("undefined"))
            {
                return false;
            }

            var documentState = (string)_javaScriptExecutor.ExecuteScript("return document.readyState");
            if (!documentState.Equals("complete"))
            {
                return false;
            }

            var jQueryActive = (long)_javaScriptExecutor.ExecuteScript("return jQuery.active");
            if (jQueryActive != 0)
            {
                return false;
            }

            return true;
        }
    }
}